File: SkyLine.py
"""
Problem:
https://leetcode.com/problems/the-skyline-problem/description/

Each building can be represented as an extent at a given height. The output
is the highest extent covering each extent start, and the second highest extent
covering each extent end.

The solution here is to build a set of points representing each extent
start/end. Iterate over the points in ascending order. For each point, add all
extents that cover that point (excluding the extent end) to a priority queue.
Then from the queue extract the first extent that covers that point. This is the
highest extent that covers that point. Output the coordinates if the height is
different from the previous height.

The runtime is O(nlogn) for sorting and priority queue. O(n) memory.
"""

import queue
import unittest

from typing import List

class Extent:
	def __init__(self, l, r, h):
		self.l = l
		self.r = r
		self.h = h

	def __lt__(self, other):
		if self.l == other.l:
			if self.h == other.h:
				return self.r < other.r
			return self.h > other.h
		return self.l < other.l

	def __repr__(self):
		return "Extent({},{},{})".format(self.l, self.r, self.h)

class Solution:
	def getSkyline(self, buildings: List[tuple[int, int, int]]) -> List[tuple[int, int]]:
		if len(buildings) == 0:
			return []

		points = set()
		for b in buildings:
			points.add(b[0])
			points.add(b[1])
		points = sorted(points)

		extents = sorted(map(lambda b: Extent(b[0], b[1], b[2]), buildings))

		q = queue.PriorityQueue()
		q.put((0, Extent(min(points), max(points), 0)))

		res = []
		i = 0
		for p in points:
			while i < len(extents) and extents[i].l <= p:
				q.put((-extents[i].h, extents[i]))
				i += 1
			while not q.empty():
				e = q.get()[1]
				if e.l <= p and e.r > p:
					if len(res) == 0 or res[-1][1] != e.h:
						res.append((p, e.h))
					q.put((-e.h, e))
					break
		res.append((max(points), 0))
		return res

class TestSkyline(unittest.TestCase):

	def test1(self):
		buildings = [[2,9,10],[3,7,15],[5,12,12],[15,20,10],[19,24,8]]
		expectedOutput = [(2,10),(3,15),(7,12),(12,0),(15,10),(20,8),(24,0)]
		actualOutput = Solution().getSkyline(buildings)
		self.assertEqual(actualOutput, expectedOutput)

	def test2(self):
		buildings = [[0,2,3],[2,5,3]]
		expectedOutput = [(0,3),(5,0)]
		actualOutput = Solution().getSkyline(buildings)
		self.assertEqual(actualOutput, expectedOutput)

	def testEmpty(self):
		self.assertEqual(Solution().getSkyline([]), [])

	def testSingle(self):
		self.assertEqual(Solution().getSkyline([(1,2,3)]), [(1,3),(2,0)])

	def testContained(self):
		buildings = [(1, 10, 5), (3, 6, 3)]
		output = [(1, 5), (10, 0)]
		self.assertEqual(Solution().getSkyline(buildings), output)

if __name__ == '__main__':
	unittest.main()

File: skyline.rs
// See Skyline.py for a more detailed description.

use std::collections::{ BinaryHeap, BTreeSet };

#[derive(Debug, Eq, PartialEq)]
struct Extent {
    l: i32,
    r: i32,
    h: i32,
}

impl Extent {
    fn new(l: i32, r: i32, h: i32) -> Self {
	Self { l, r, h }
    }
}

impl Ord for Extent {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
	if self.h == other.h {
	    if self.l == other.l {
		return self.r.cmp(&other.r);
	    }
	    return self.l.cmp(&other.l);
	}
	return self.h.cmp(&other.h);
    }
}

impl PartialOrd for Extent {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

fn lhr_cmp(this: &Extent, that: &Extent) -> std::cmp::Ordering {
    if this.l == that.l {
	if this.h == that.h {
	    return this.r.cmp(&that.r);
	}
	return this.h.cmp(&that.h);
    }
    return this.l.cmp(&that.l);
}

fn skyline(buildings: &Vec<(i32, i32, i32)>) -> Vec<(i32, i32)> {
    if buildings.is_empty() {
	return Vec::new();
    }

    let mut extents = buildings
	.iter()
	.map(|b| Extent::new(b.0, b.1, b.2)).collect::<Vec<Extent>>();
    extents.sort_by(lhr_cmp);

    let mut points_set = BTreeSet::new();
    for e in extents.iter() {
	points_set.insert(e.l);
	points_set.insert(e.r);
    }

    let mut q = BinaryHeap::new();
    let base_extent = Extent::new(
	*points_set.first().unwrap(),
	*points_set.last().unwrap(),
	0,
    );
    q.push(&base_extent);

    let mut res: Vec<(i32, i32)> = Vec::new();
    let mut i = 0;
    for p in points_set.iter() {
	while i < extents.len() && extents[i].l <= *p {
	    q.push(&extents[i]);
	    i += 1;
	}
	while !q.is_empty() {
	    if let Some(e) = q.pop() {
		if e.r > *p {
		    if res.is_empty() || res.last().unwrap().1 != e.h {
			res.push((*p, e.h));
		    }
		    q.push(e);
		    break;
		}
	    }
	}
    }
    res.push((*points_set.last().unwrap(), 0));
    res
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test1() {
	let buildings = vec![(2,9,10),(3,7,15),(5,12,12),(15,20,10),(19,24,8)];
	let expected_output = vec![(2,10),(3,15),(7,12),(12,0),(15,10),(20,8),(24,0)];
	let actual_output = skyline(&buildings);
	assert_eq!(expected_output, actual_output);
    }

    #[test]
    fn test2() {
	let buildings = vec![(0,2,3),(2,5,3)];
	let expected_output = vec![(0,3),(5,0)];
	let actual_output = skyline(&buildings);
	assert_eq!(expected_output, actual_output);
    }

    #[test]
    fn test_empty() {
	assert_eq!(skyline(&Vec::new()), Vec::new());
    }

    #[test]
    fn test_single() {
	assert_eq!(skyline(&vec![(1,2,3)]), vec![(1,3),(2,0)]);
    }

    #[test]
    fn test_contained() {
	let buildings = vec![(1,10,5),(3,6,3)];
	let output = vec![(1,5),(10,0)];
	assert_eq!(skyline(&buildings), output);
    }
}
